File: cpp/teach_OCP.cpp
#include <iostream>
#include <string>

// Open-Closed Principle - OCP
// Opened for extension. Closed for modification


//
//  Interface of Human
//
class InterfaceHuman{
public:
    friend std::ostream& operator<< (std::ostream& out, const InterfaceHuman& iHuman);
    
    virtual ~InterfaceHuman() {}
    
    virtual void go() = 0;  // {}   // pass statement from Python == empty block statement {} in C++
                                    // but that is not pure virtual
};

std::ostream& operator<< (std::ostream& out, const InterfaceHuman& iHuman){
    return out;
}

//
//  Human
//
class Human: public InterfaceHuman{
public:
    friend std::ostream& operator<< (std::ostream& out, const Human& human);
    
    explicit Human(std::string inName, int inAge, float inWeight):
        m_name(inName),
        m_age(inAge),
        m_weight(inWeight)
    {}

    void go(){
        //InterfaceHuman::go(); // Error while Interface method is pure virtual !!!
        std::cout << "Human " << m_name << " GO!" << std::endl;
    }

    std::string getName() const{
        return m_name;
    }

private:
    std::string m_name;
    int         m_age;
    float       m_weight;
};

std::ostream& operator<< (std::ostream& out, const Human& human){
    out << "Human: " 
    << human.m_name << " | "
    << human.m_age << " | "
    << human.m_weight;

    return out;
}


//
//  Employee
//
class Employee: public Human{
public:
    friend std::ostream& operator<< (std::ostream& out, const Employee& employee);

    explicit Employee(std::string inName, int inAge, float inWeight, int inIDcard):
        Human(inName, inAge, inWeight),
        m_IDcard(inIDcard)
    {}

    void go(){
        std::cout << "Employee " 
        << getName() << " | " 
        << getIDcard() << " GOO!!" << std::endl;
    }

    int getIDcard() const{
        return m_IDcard;
    }

private:
    int m_IDcard;
};

std::ostream& operator<< (std::ostream& out, const Employee& employee){
    out << static_cast<Human>(employee) << " | " << employee.m_IDcard;

    return out;
}


int main(){
    Human human("Lev Grodzitskiy", 32, 111.3);
    std::cout << human << std::endl;
    human.go();

    Employee employee("Lev Grodzitskiy", 32, 111.3, 45183);
    std::cout << employee << std::endl;
    employee.go();

    std::cout << "Done!" << std::endl;

    return 0;
}

File: python/teach_OCP.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Open-Closed Principle - OCP
# Opened for expansion. Closed for modification

class InterfaceHuman(object):
    def __str__(self):
        pass

    def go(self):
        pass


class Human(InterfaceHuman):
    def __init__(self, inName: str, inAge: int, inWeight: float):
        self.__name = inName
        self.__age = inAge
        self.__weight = inWeight

    def __str__(self):
        InterfaceHuman.__str__(self)
        return f"Human: {self.__name} | {self.__age} | {self.__weight}"

    def go(self):
        InterfaceHuman.go(self)
        print(f"Human {self.__name} GO!")

    @property
    def name(self):
        return self.__name


class Employee(Human):
    def __init__(self, inName: str, inAge: int, inWeight: float, inIdCard: int):
        super().__init__(inName, inAge, inWeight)
        # Human.__init__(self, inName, inAge, inWeight)
        self.__IdCard = inIdCard

    def __str__(self):
        return Human.__str__(self) + f"| {self.__IdCard}"

    def go(self):
        print(f"Employee {self.name} | {self.__IdCard} GOO!!")

    @property
    def IdCard(self):
        return self.__IdCard


if __name__ == "__main__":
    ihuman = InterfaceHuman()
    ihuman.go()

    human = Human("Lev Grodzitskiy", 32, 111.3)
    print(human)
    human.go()

    employee = Employee("Lev Grodzitskiy", 32, 111.3, 45183)
    print(employee)
    employee.go()

